### PR TITLE
StateManager interface

### PR DESCRIPTION
**Pull request by s1na**
So I've added an interface for the StateManager. Had to make a few other changes:

- The precompile activation logic was using stateManager._trie directly, replaced that with `putAccount`.
- Also the blake2f precompile wasn't activated because the addresses were hardcoded. Fixed now.
- `_clearOriginalStorageCache` was private, made it public. The issue is that `StateManager` is persistent state, but the originalStorageCache should only persist for the duration of one tx.

Should I move the docs to the interface or leave them in the implementation? Also please feel free to suggest better names for `DefaultStateManager` :D

### CODE REVIEW

**Comment by s1na** [comment]
Rebased master, that should hopefully fix the codecov issue? also should I rebuild the docs in the PR or is that done at release?

**Comment by evertonfraga** [comment]
@s1na I got started late today, but it seems Codecov issue is gone.

image

And LGTM finally finished its report.

File: packages/vm/lib/state/interface.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import Account from 'ethereumjs-account'
+
+/**
+ * Storage values of an account
+ */
+export interface StorageDump {
+  [key: string]: string
+}
+
+export interface StateManager {
```
Comment (ryanio) [inline packages/vm/lib/state/interface.ts:10]: 👌

**Comment by ryanio** [comment]
> also should I rebuild the docs in the PR or is that done at release?

@s1na you can rebuild, I usually do it as a last commit. it would be a nice improvement to ensure docs automatically rebuild prior to release.

**Comment by evertonfraga** [comment]
yeah, IMO having the docs generated for each PR change ensures the docs reflect exactly what’s on master branch. 

Thinking of people that land on the project to experiment, having tidy docs can only help them =)

**Comment by s1na** [comment]
Note to self: `stateManager._touched` is also cleared at the end of the tx, but by checking on `commit()` whether the lists of checkpoints is emtpy. Same trick can probably be used for the storage original cache.

**Comment by evertonfraga** [comment]
> Also the blake2f precompile wasn't activated because the addresses were hardcoded. Fixed now.

Hi @s1na, before I move on with rebase & merge, I wanted to understand the aforementioned change. Pardon my ignorance (😄), but where exactly is this issue fixed, and what's the impact of the current state (eg: blake2f not operational)?

**Comment by s1na** [comment]
Hey @evertonfraga, no it only fixes a small edge case. Contracts that call blake2f will work. For context on what `activatePrecompiles` does see #575. In this case the new precompile wasn't being "activated" (confusing terminology) as you can see here https://github.com/ethereumjs/ethereumjs-vm/pull/763/files#diff-1c815e0b4a51825344018a6a6c522733L150 (note the range `i = 0; i < 8; i++` which doesn't cover `i = 9` which is the blake2f address).

Sorry I haven't the PR yet, my working tree for this branch is dirty because I just started working on the flat db without creating a new branch... will update now with re-built docs and dropping the `clearOriginalStorageCache` method.

**Comment by evertonfraga** [comment]
@s1na thanks for the explanation! Yes, the `activated precompiles` terminology have caught me. I have seen that activation behavior before, just didn't know its name or where it was represented in the code. Thanks for pointing it out.

I will now rebase and merge this PR. Sounds good?

**Comment by s1na** [comment]
Update: I made `clearOriginalStorageCache` private again (so its not part of the interface), and removed it from `runTx`. It's now run implicitly when the highest-level checkpoint terminates. This happens when a tx has finished processing (but could happen in other cases too, hope its not confusing).

Oh and I'm not sure if this PR should be considered a breaking change due to `StateManager` being renamed to `DefaultStateManager`?

Also tried rebuilding docs but couldn't get it to work, was constantly getting an error.

If everything looks good the PR can be merged.

**Comment by evertonfraga** [comment]
> Update: I made clearOriginalStorageCache private again (so its not part of the interface), and removed it from runTx. 
Alright!

> Oh and I'm not sure if this PR should be considered a breaking change due to StateManager being renamed to DefaultStateManager?
The VM will be bumped to v5.0.0 anyway, so we shouldn't worry much about it. I will, of course, make a note about this change in the release notes.

> Also tried rebuilding docs but couldn't get it to work, was constantly getting an error.

I'll handle it!

**Comment by holgerd77** [comment]
Short note: did a branch update here, hope that was ok, so that it might be merge-ready once CI pass.

**Comment by evertonfraga** [comment]
Thanks @holgerd77!

**Pull request merged**